csharp/ql/src/Likely Bugs/MishandlingJapaneseEra.ql

```diff
@@ -0,0 +1,60 @@
+/**
+ * @name MishandlingJapaneseEraStartDate
+ * @description Japanese Era should be handled with built-in JapaneseCalendar class. Aviod hard-coding Japanese era start dates and names. 
+ * @id cs/mishandling-japanese-era
+ * @kind problem
+ * @problem.severity warning
+ * @precision medium
+ * @tags reliability
+ *       japanese-era
+ */
+
+import csharp
+
+predicate isExactEraStartDateCreation(ObjectCreation cr) {
+  (cr.getType().hasQualifiedName("System.DateTime") or cr.getType().hasQualifiedName("System.DateTimeOffset"))
+  and  
+    cr.getArgument(0).getValue() = "1989" and
+    cr.getArgument(1).getValue() = "1" and
+    cr.getArgument(2).getValue() = "8"
+}
+
+predicate isDateFromJapaneseCalendarToDateTime(MethodCall mc) {
+  (mc.getQualifier().getType().hasQualifiedName("System.Globalization.JapaneseCalendar") or mc.getQualifier().getType().hasQualifiedName("System.Globalization.JapaneseLunisolarCalendar"))
+  and 
+    mc.getTarget().hasName("ToDateTime") and
+    mc.getArgument(0).getValue() != "" and
+    (mc.getNumberOfArguments() = 7 or // implicitly current era
+     mc.getNumberOfArguments() = 8 and
+     mc.getArgument(7).getValue() = "0") // explicitly current era
+}
+
+predicate isDateFromJapaneseCalendarCreation(ObjectCreation cr) {
+  (cr.getType().hasQualifiedName("System.DateTime") or cr.getType().hasQualifiedName("System.DateTimeOffset")) and
+  (cr.getArgumentForName("calendar").getType().hasQualifiedName("System.Globalization.JapaneseCalendar") or cr.getArgumentForName("calendar").getType().hasQualifiedName("System.Globalization.JapaneseLunisolarCalendar"))
+  and  
+    cr.getArgumentForName("year").getValue() != ""
+}
+
+predicate inEraArrayCreation(ArrayInitializer ai) {
+  ai.getElement(0).getValue() = "1867" and
+  ai.getElement(1).getValue() = "1911" and
+  ai.getElement(2).getValue() = "1925" and
+  ai.getElement(3).getValue() = "1988"
+}
+
+predicate isEraCollectionCreation(CollectionInitializer cs) {
+  cs.getElementInitializer(0).getValue() = "1867" and
+  cs.getElementInitializer(1).getValue() = "1911" and
+  cs.getElementInitializer(2).getValue() = "1925" and
+  cs.getElementInitializer(3).getValue() = "1988"
```

Comment:
2019?

Reply:
Probably worth factoring out the different eras as a separate predicate and then refer to that throughout:
```
/**
 * Holds if `year`, `month`, and `day` specify the start of a new era
 * (see https://simple.wikipedia.org/wiki/List_of_Japanese_eras).
 */
predicate isEraStart(int year, int month, int day) {
  // Add more cases ...
  year = 1989 and month = 2 and day = 1
  or
  year = 2019 and month = 5 and day = 1
}

predicate inEraArrayCreation(ArrayInitializer ai) {
  isEraStart(ai.getElement(0).getValue().toInt(), _, _)
}

predicate isEraCollectionCreation(CollectionInitializer cs) {
  isEraStart(cs.getElementInitializer(0).getValue().toInt(), _, _)
}
```